OpenRA.Game/Widgets/WidgetUtils.cs

```diff
@@ -203,29 +203,29 @@ public static string WrapText(string text, int width, SpriteFont font)
 				for (var i = 0; i < lines.Count; i++)
 				{
 					var line = lines[i];
-					var m = font.Measure(line);
-
-					if (m.X <= width)
+					if (font.Measure(line).X <= width)
 						continue;
 
-					var bestSpaceIndex = -1;
-					var start = line.Length - 1;
-
-					while (m.X > width)
+					// Scan forwards until we find the last word that fits
+					// This guarantees a small bound on the amount of string we need to search before a linebreak
+					var start = 0;
+					while (true)
 					{
-						var spaceIndex = line.LastIndexOf(' ', start);
+						var spaceIndex = line.IndexOf(' ', start);
 						if (spaceIndex == -1)
 							break;
-						bestSpaceIndex = spaceIndex;
 
-						start = spaceIndex - 1;
-						m = font.Measure(line.Substring(0, spaceIndex));
+						var fragmentWidth = font.Measure(line.Substring(0, spaceIndex)).X;
```

Comment:
This still has quadratic behaviour, you could maintain the fragment length so far and just compute the increments to get back to linear time.

Reply:
I wanted to avoid a potenial can of worms with kerning and ligatures – i'm not sure whether they interact with inter-word spaces, but it seems plausible that they could.  The current approach is indeed still quadratic, but the term being squared is now bounded to &lesssim; 20.

Reply:
Given that `Measure` just blindly sums up characters widths, I think that's a can of worms we already opened and are wilfully ignoring (unless there's some magic I'm missing).

Reply:
I'd like to think that we could fix that one day, but I guess we can always undo the change here if that day ever comes.

Reply:
This still doesn't seem to work.  Consider the following test case:
```csharp

var font = Game.Renderer.Fonts["Regular"];
var testString = "Pellentesque dapibus interdum odio, ut rhoncus arcu malesuada non. Fusce bibendum, tortor";
var testFragments = new[] {
	"Pellentesque ",
	"dapibus ",
	"interdum ",
	"odio, ",
	"ut ",
	"rhoncus ",
	"arcu ",
	"malesuada ",
	"non. ",
	"Fusce ",
	"bibendum, ",
	"tortor"
};

Console.WriteLine("single: {0} aggregate: {1}", font.Measure(testString).X, testFragments.Select(x => font.Measure(x).X).Sum());
```
> single: 588 aggregate: 591

Am I missing something obvious?

Reply:
> single: 592 aggregate: 592

Works on my machine? Happy to defer this to another time though.